packages/aws-cdk-lib/aws-cloudfront-origins/lib/s3-origin.ts

```diff
@@ -28,12 +79,25 @@ export class S3Origin implements cloudfront.IOrigin {
   private readonly origin: cloudfront.IOrigin;
 
   constructor(bucket: s3.IBucket, props: S3OriginProps = {}) {
-    this.origin = bucket.isWebsite ?
-      new HttpOrigin(bucket.bucketWebsiteDomainName, {
+    if (props.originAccessControl && props.originAccessIdentity) {
+      throw new Error('Only one of originAccessControl or originAccessIdentity can be specified for an origin.');
+    }
+
+    if (bucket.isWebsite) {
+      this.origin = new HttpOrigin(bucket.bucketWebsiteDomainName, {
         protocolPolicy: cloudfront.OriginProtocolPolicy.HTTP_ONLY, // S3 only supports HTTP for website buckets
         ...props,
-      }) :
-      new S3BucketOrigin(bucket, props);
+      });
+    } else if (props.originAccessControl) {
+      this.origin = S3BucketOrigin.withAccessControl(bucket, props);
+    } else if (props.originAccessIdentity) {
+      this.origin = S3BucketOrigin.withAccessIdentity(bucket, props);
+    } else {
+      this.origin = FeatureFlags.of(bucket.stack)
```

Comment:
Should you check current stack or bucket stack for cross stack usage? IMO ideally it doesn't make a difference since the Feature flag is app-level not stack-level. Just want to confirm.